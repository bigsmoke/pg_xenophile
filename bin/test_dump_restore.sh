#!/bin/bash

SCRIPT_NAME=$(basename "$0")
PG_BIN_DIR="$(pg_config --bindir)"

usage() {
    cat <<EOF
Usage:
    $SCRIPT_NAME [options] --extension <extension_name> --psql-script-file <file> --out-file <file> --expected-out-file <file>
    $SCRIPT_NAME --help|-h

Options:
    --keep-temp-dir
EOF
}

psql_script_file=""
expected_out_file=""
out_file=""
extension_name=""
keep_temp_dir=""
while [[ "$#" -gt 0 ]]; do
    case "$1" in
        -h|--help)
            usage
            exit 0
            ;;
        --psql-script-file)
            psql_script_file="$2"
            shift 2
            ;;
        --out-file)
            out_file="$2"
            shift 2
            ;;
        --expected-out-file)
            expected_out_file="$2"
            shift 2
            ;;
        --extension)
            extension_name="$2"
            shift 2
            ;;
        --keep-temp-dir)
            keep_temp_dir="alright"
            shift
            ;;
        --*|-*)
            echo -e "\e[31mUnrecognized option: \e[1m$1\e[0m" >&2
            usage >&2
            exit 2
            ;;
    esac
done
if [ -z "$psql_script_file" ]; then
    echo -e "\e[31mMissing param: \e[1m--psql-script-file\e[0m" >&2
    usage >&2
    exit 2
fi
if [ -z "$out_file" ]; then
    echo -e "\e[31mMissing param: \e[1m--out-file\e[0m" >&2
    usage >&2
    exit 2
fi
if [ -z "$expected_out_file" ]; then
    echo -e "\e[31mMissing param: \e[1m--expected-out-file\e[0m" >&2
    usage >&2
    exit 2
fi
if [ -z "$extension_name" ]; then
    echo -e "\e[31mMissing param: \e[1m--extension\e[0m" >&2
    usage >&2
    exit 2
fi

cleanup() {
    if [ -n "$pg_pid" ]; then
        kill "$pg_pid"
    fi

    if [ -n "$tmp_dir" ]; then
        if [ -n "$keep_temp_dir" ]; then
            echo -e "Temp. dir. preserved: \e[1m$tmp_dir\e[22m"
        else
            rm -rf "$tmp_dir"
        fi
    fi
}
trap cleanup exit

tmp_dir=$(mktemp --directory)
dump_file="$tmp_dir/dump"
roles_dump_file="$tmp_dir/global.dump"

export PGDATA="$tmp_dir/data"
export PGHOST="$tmp_dir"
export PGDATABASE="test_dump_restore"
export PGUSER="wortel"
export PGVERSION="15"
OID_NOISE_DB_NAME="test_dump_restore_oid_noise"

start_pg_cluster() {
    "$PG_BIN_DIR/initdb" \
            --auth trust \
            --username "$PGUSER" \
            --pwfile <(echo ondergronds) >/dev/null 2>&1 || exit 4

    "$PG_BIN_DIR/postgres" -k "$PGHOST" -h "" 2>/dev/null &
    pg_pid=$!

    # Wait for the daemon to come online.
    while true; do
        if psql -c "select true" postgres >/dev/null; then
            break
        fi
        sleep 0.1
    done
}
start_pg_cluster

mkdir -p $(dirname "$out_file")

echo "-- createdb" > "$out_file"
$PG_BIN_DIR/createdb || exit 5

echo "-- psql -f '$psql_script_file' -v 'extension_name=$extension_name' -v 'test_stage=pre-dump'" >> "$out_file"
$PG_BIN_DIR/psql \
    -f "$psql_script_file" \
    -v "extension_name=$extension_name" \
    -v "test_stage=pre-dump" \
    >> "$out_file" 2>&1 || exit 5

echo "-- pg_dumpall --roles-only" >> "$out_file"
$PG_BIN_DIR/pg_dumpall --roles-only --file "$roles_dump_file" >> "$out_file" 2>&1 || exit 5

echo "-- pg_dump --format=custom --file <dump_file>" >> "$out_file"
$PG_BIN_DIR/pg_dump --format=custom --file "$dump_file" >> "$out_file" 2>&1 || exit 5

echo "-- kill -9 <pg_pid>" >> "$out_file"
kill "$pg_pid"
echo "-- rm -r \$PGDATA" >> "$out_file"
rm -r "$PGDATA"

start_pg_cluster

echo "-- psql postgres -c '\\set ON_ERROR_STOP' -f <roles_dump_file>" >> "$out_file"
$PG_BIN_DIR/psql postgres -c '\set ON_ERROR_STOP' \
    -f <(grep -v "CREATE ROLE $PGUSER" "$roles_dump_file") \
    >> "$out_file" 2>&1 \
    || exit 5

echo "-- createdb '$OID_NOISE_DB_NAME'" >> "$out_file"
$PG_BIN_DIR/createdb "$OID_NOISE_DB_NAME" >> "$out_file" 2>&1 || exit 5

echo "-- psql -f '$psql_script_file' -v 'extension_name=$extension_name' -v 'test_stage=pre-restore'" >> "$out_file"
$PG_BIN_DIR/psql \
    -f "$psql_script_file" \
    -v "extension_name=$extension_name" \
    -v "test_stage=pre-restore" \
    "$OID_NOISE_DB_NAME" \
    >> "$out_file" 2>&1 || exit 5

echo "-- pg_restore --create --dbname postgres <dump_file>" >> "$out_file"
$PG_BIN_DIR/pg_restore --exit-on-error --create --dbname postgres "$dump_file" >> "$out_file" 2>&1 || exit 5

echo "-- psql -f '$psql_script_file' -v 'extension_name=$extension_name' -v 'test_stage=post-restore'" >> "$out_file"
$PG_BIN_DIR/psql \
    -f "$psql_script_file" \
    -v "extension_name=$extension_name" \
    -v "test_stage=post-restore" \
    >> "$out_file" 2>&1 || exit 5

if ! diff "$expected_out_file" "$out_file" >/dev/null; then
    echo "Expected output was not the same as the actual output:"
    diff "$expected_out_file" "$out_file"
    exit 6
fi

# vim: set expandtab shiftwidth=4 tabstop=4:
