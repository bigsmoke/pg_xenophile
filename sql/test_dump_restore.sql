\o /dev/null
select  not :{?test_stage} as test_stage_missing
        ,not :{?extension_name} as extension_name_missing;
\o
\gset
\if :test_stage_missing
    \warn 'Missing `:test_stage` variable.'
    \quit
\endif
\if :extension_name_missing
    \warn 'Missing `:extension_name` variable.'
    \quit
\endif
\o /dev/null
select  :'test_stage' = 'pre-dump' as in_pre_dump_stage
        ,:'test_stage' = 'pre-restore' as in_pre_restore_stage;
\o
\gset

\set SHOW_CONTEXT 'errors'

\if :in_pre_restore_stage
    -- Let's generate some noise to offset the OIDs, to ensure that we're not relying on OIDs remaining the
    -- same between the moment of `pg_dump` and the moment of `pg_restore`.
    do $$
    declare
        _i int;
    begin
        for
            _i
        in select
            s.i
        from
            generate_series(1, 10) as s(i)
        loop
            execute format('CREATE TABLE test_dump_restore__oid_noise__tbl_%s (a int)', _i);
            execute format('CREATE TYPE test_dump_restore__oid_noise__rec_%s AS (a int, b int)', _i);
            execute format(
                'CREATE FUNCTION test_dump_restore__oid_noise__func_%s() RETURNS int RETURN 1'
                ,_i
            );
        end loop;
    end;
    $$;

    \quit
\endif

\if :in_pre_dump_stage
    create extension pg_xenophile with cascade;
\endif

call xeno.test_dump_restore__l10n_table(:'test_stage'::text);
