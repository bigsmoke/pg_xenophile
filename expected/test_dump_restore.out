-- createdb
-- psql -f 'sql/test_dump_restore.sql' -v 'extension_name=pg_xenophile' -v 'test_stage=pre-dump'
psql:sql/test_dump_restore.sql:51: NOTICE:  installing required extension "hstore"
CREATE EXTENSION
CALL
-- pg_dumpall --roles-only
-- pg_dump --format=custom --file <dump_file>
-- kill -9 <pg_pid>
-- rm -r $PGDATA
-- psql postgres -c '\set ON_ERROR_STOP' -f <roles_dump_file>
SET
SET
SET
ALTER ROLE
-- createdb 'test_dump_restore_oid_noise'
-- psql -f 'sql/test_dump_restore.sql' -v 'extension_name=pg_xenophile' -v 'test_stage=pre-restore'
DO
-- pg_restore --create --dbname postgres <dump_file>
-- psql -f 'sql/test_dump_restore.sql' -v 'extension_name=pg_xenophile' -v 'test_stage=post-restore'
CALL
